src/rgw/driver/d4n/d4n_policy.cc

```diff
@@ -243,6 +243,39 @@ asio::awaitable<void> LFUDAPolicy::redis_sync(const DoutPrefixProvider* dpp, opt
   }
 }
 
+// Changes state to INVALID for dirty objects. An INVALID state indicates that a delete request has been
+// issued on an object and it must be deleted rather than written to the backend. This lazy deletion occurs
+// in the Cleaning method and prevents data races during concurrent requests. This method returns "false"
+// if the state has not been set to INVALID, and "true" if it has.
+bool LFUDAPolicy::invalidate_dirty_object(const DoutPrefixProvider* dpp, std::string& key) {
+  std::unique_lock<std::mutex> l(lfuda_cleaning_lock);
+
+  if (o_entries_map.empty())
+    return false;
+  
+  auto p = o_entries_map.find(key);
+  if (p == o_entries_map.end()) {
+    ldpp_dout(dpp, 10) << "LFUDAPolicy::" << __func__ << "(): key=" << key << " not found" << dendl;
+    return false;
+  }
+
+  if (p->second.second == State::INIT) {
+    ldpp_dout(dpp, 10) << "LFUDAPolicy::" << __func__ << "(): Setting State::INVALID for key=" << key << dendl;
+    p->second.second = State::INVALID;
+    int ret = cacheDriver->set_attr(dpp, "D#" + key, RGW_CACHE_ATTR_INVALID, "1", y);
+    if (ret < 0) {
+      ldpp_dout(dpp, 0) << "LFUDAPolicy::" << __func__ << "(): Failed to set xattr, ret=" << ret << dendl;
+      return false;
+    }
+  
+    return true;
+  } else if (p->second.second == State::IN_PROGRESS) {
+    state_cond.wait(l, [this, &p]{ return (p->second.second == State::DONE); });
```

Comment:
Here you are accessing the state as part of the map entry and the map entry gets erased in `erase_dirty_object`. P.S: in the predicate for condition wait you need to first get the map entry and then check its state -  and this will become a problem because it would have been erased in `erase_dirty_object`,  and you need to fetch the state here otherwise you will always have a stale value of state set to IN_PROGRESS. Do you understand the problem now?

Reply:
And now when I think about it when you fetch the map entry afresh in the predicate, it won't be found.

Reply:
Yes, that initial search for the entry would be an issue. As a result, the correct condition is to wait for that entry to no longer exists in the dirty heap. Is my understanding correct?

Reply:
Yes, and remove DONE state